### PR TITLE
Is Jetpack Fast yet? (proof of concept)

### PR DESCRIPTION
**Pull request by simison**
Hey! :wave:

I just wanted to show you what I was doing for your app while studying Jetpack build process, Webpack and your app all at the same time. (I'm new to Webpack)

Since most of the stuff in Jetpack [doesn't go through Webpack](https://github.com/Automattic/jetpack/blob/b2523e5e39ca95c639acfaca88932df2421c0b30/gulpfile.babel.js#L110-L126), this is kinda useless for now.

Jetpack also has a ton of other speed considerations to measure; JS bundle size isn't nearly as important as it's for Calypso. Other measurements could be easy enough to add to the same app, though!

That said, I might've added something you want to add to ICFY. Some bug fixes, docker-compose setup etc. Up to you!

FYI @oskosk

image

### CODE REVIEW

File: server/db.js [hunk 1]
```diff
@@ -10,7 +10,11 @@ exports.getPush = sha =>
 		.select()
 		.where('sha', sha);
 
-exports.insertPush = push => K('pushes').insert(push);
+exports.insertPush = (push) => {
+	// Convert '2018-03-27T22:14:19Z' to '2018-03-27 22:14:19'
+	push.created_at = new Date(push.created_at).toISOString().slice(0, 19).replace('T', ' ');
```
[the remark this answers is not in the document]
Comment (simison) [reply server/db.js]: It couldn't insert the push due MySQL complaining about wrong format for the `datetime` field.

Not sure why it works for you and didn't work for me, though. 🤔

[the remark this answers is not in the document]
Comment (jsnajdr) [reply server/db.js]: In my database, the `pushes.created_at` column is of type TIMESTAMP, not DATETIME:
screen shot 2018-03-29 at 12 30 56

Maybe that's the reason? Also, passing `Date` object, i.e., `push.created_at = new Date(push.created_at)` might work as well.

**Comment by simison** [comment]
>I love the idea of making the tool support multiple repositories 

<3 totally. This was a good experiment that shows it wouldn't require a ton of work to get started. To have even just most simple multi-app version running would inspire others to improve it.

We'd need just some configs and some abstraction for the builder. Initially, we could even just run separate DBs/servers/clients for each app under the same domain. MVP would be just making sure everything works with the same codebase.

Personally, I'd be interested in exploring some other ways to collect stats as well. It would be quite straightforward to launch new Jetpack site with https://jurassic.ninja/ or [Jetpack-Docker](https://github.com/Automattic/jetpack/tree/master/docker) and run something like https://github.com/GoogleChrome/lighthouse in it.

BTW setting the app up was quite a bit of trial-error, so even just having better readme would make this more accessible to others. ;-)

**Comment by simison** [comment]
I'm also getting this occasionally:

```
server_1  | [2] Failed to analyze bundle stats:
server_1  | [2] { Error: ER_DUP_ENTRY: Duplicate entry 'b2523e5e39ca95c639acfaca88932df2421c0b30' for key 'chunk_groups_sha_unique'
server_1  | [2]     at Query.Sequence._packetToError (/srv/server/node_modules/mysql/lib/protocol/sequences/Sequence.js:52:14)
server_1  | [2]     at Query.ErrorPacket (/srv/server/node_modules/mysql/lib/protocol/sequences/Query.js:77:18)
server_1  | [2]     at Protocol._parsePacket (/srv/server/node_modules/mysql/lib/protocol/Protocol.js:279:23)
server_1  | [2]     at Parser.write (/srv/server/node_modules/mysql/lib/protocol/Parser.js:76:12)
server_1  | [2]     at Protocol.write (/srv/server/node_modules/mysql/lib/protocol/Protocol.js:39:16)
server_1  | [2]     at Socket.<anonymous> (/srv/server/node_modules/mysql/lib/Connection.js:103:28)
server_1  | [2]     at emitOne (events.js:116:13)
server_1  | [2]     at Socket.emit (events.js:211:7)
server_1  | [2]     at addChunk (_stream_readable.js:263:12)
server_1  | [2]     at readableAddChunk (_stream_readable.js:250:11)
server_1  | [2]     at Socket.Readable.push (_stream_readable.js:208:10)
server_1  | [2]     at TCP.onread (net.js:594:20)
server_1  | [2]     --------------------
server_1  | [2]     at Protocol._enqueue (/srv/server/node_modules/mysql/lib/protocol/Protocol.js:145:48)
server_1  | [2]     at Connection.query (/srv/server/node_modules/mysql/lib/Connection.js:208:25)
server_1  | [2]     at /srv/server/node_modules/knex/lib/dialects/mysql/index.js:152:18
server_1  | [2]     at Promise._execute (/srv/server/node_modules/knex/node_modules/bluebird/js/release/debuggability.js:303:9)
server_1  | [2]     at Promise._resolveFromExecutor (/srv/server/node_modules/knex/node_modules/bluebird/js/release/promise.js:483:18)
server_1  | [2]     at new Promise (/srv/server/node_modules/knex/node_modules/bluebird/js/release/promise.js:79:10)
server_1  | [2]     at Client_MySQL._query (/srv/server/node_modules/knex/lib/dialects/mysql/index.js:146:12)
server_1  | [2]     at Client_MySQL.query (/srv/server/node_modules/knex/lib/client.js:197:17)
server_1  | [2]     at Runner.<anonymous> (/srv/server/node_modules/knex/lib/runner.js:146:36)
server_1  | [2]     at Runner.tryCatcher (/srv/server/node_modules/knex/node_modules/bluebird/js/release/util.js:16:23)
server_1  | [2]     at Runner.query (/srv/server/node_modules/knex/node_modules/bluebird/js/release/method.js:15:34)
server_1  | [2]     at /srv/server/node_modules/knex/lib/runner.js:65:21
server_1  | [2]     at tryCatcher (/srv/server/node_modules/knex/node_modules/bluebird/js/release/util.js:16:23)
server_1  | [2]     at /srv/server/node_modules/knex/node_modules/bluebird/js/release/using.js:185:26
server_1  | [2]     at tryCatcher (/srv/server/node_modules/knex/node_modules/bluebird/js/release/util.js:16:23)
server_1  | [2]     at Promise._settlePromiseFromHandler (/srv/server/node_modules/knex/node_modules/bluebird/js/release/promise.js:512:31)
server_1  | [2]   code: 'ER_DUP_ENTRY',
server_1  | [2]   errno: 1062,
server_1  | [2]   sqlMessage: 'Duplicate entry \'b2523e5e39ca95c639acfaca88932df2421c0b30\' for key \'chunk_groups_sha_unique\'',
server_1  | [2]   sqlState: '23000',
server_1  | [2]   index: 0,
server_1  | [2]   sql: 'insert into `chunk_groups` (`chunk`, `sha`, `sibling`) values (\'build\', \'b2523e5e39ca95c639acfaca88932df2421c0b30\', \'manifest\'), (\'build\', \'b2523e5e39ca95c639acfaca88932df2421c0b30\', \'vendor\')' }
```

...that's why added that try-catch so that the server wouldn't stall on those.

With Docker, I could just tarball my folder, send it over you and you could run exact clone of what I have now (including the database). ;-)

File: server/db.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (simison) [reply server/db.js]: Hmm, same same:
image

To create these tables I simply ran `npx knex migrate:latest`.

Here's the error I get without that line:
```
server_1  | [1] Error while checking for new pushes: { Error: ER_TRUNCATED_WRONG_VALUE: Incorrect datetime value: '2018-03-29T07:29:40Z' for column 'created_at' at row 1
server_1  | [1]     at Query.Sequence._packetToError (/srv/server/node_modules/mysql/lib/protocol/sequences/Sequence.js:52:14)
server_1  | [1]     at Query.ErrorPacket (/srv/server/node_modules/mysql/lib/protocol/sequences/Query.js:77:18)
server_1  | [1]     at Protocol._parsePacket (/srv/server/node_modules/mysql/lib/protocol/Protocol.js:279:23)
server_1  | [1]     at Parser.write (/srv/server/node_modules/mysql/lib/protocol/Parser.js:76:12)
server_1  | [1]     at Protocol.write (/srv/server/node_modules/mysql/lib/protocol/Protocol.js:39:16)
server_1  | [1]     at Socket.<anonymous> (/srv/server/node_modules/mysql/lib/Connection.js:103:28)
server_1  | [1]     at emitOne (events.js:116:13)
server_1  | [1]     at Socket.emit (events.js:211:7)
server_1  | [1]     at addChunk (_stream_readable.js:263:12)
server_1  | [1]     at readableAddChunk (_stream_readable.js:250:11)
server_1  | [1]     at Socket.Readable.push (_stream_readable.js:208:10)
server_1  | [1]     at TCP.onread (net.js:594:20)
server_1  | [1]     --------------------
server_1  | [1]     at Protocol._enqueue (/srv/server/node_modules/mysql/lib/protocol/Protocol.js:145:48)
server_1  | [1]     at Connection.query (/srv/server/node_modules/mysql/lib/Connection.js:208:25)
server_1  | [1]     at /srv/server/node_modules/knex/lib/dialects/mysql/index.js:152:18
server_1  | [1]     at Promise._execute (/srv/server/node_modules/knex/node_modules/bluebird/js/release/debuggability.js:303:9)
server_1  | [1]     at Promise._resolveFromExecutor (/srv/server/node_modules/knex/node_modules/bluebird/js/release/promise.js:483:18)
server_1  | [1]     at new Promise (/srv/server/node_modules/knex/node_modules/bluebird/js/release/promise.js:79:10)
server_1  | [1]     at Client_MySQL._query (/srv/server/node_modules/knex/lib/dialects/mysql/index.js:146:12)
server_1  | [1]     at Client_MySQL.query (/srv/server/node_modules/knex/lib/client.js:197:17)
server_1  | [1]     at Runner.<anonymous> (/srv/server/node_modules/knex/lib/runner.js:146:36)
server_1  | [1]     at Runner.tryCatcher (/srv/server/node_modules/knex/node_modules/bluebird/js/release/util.js:16:23)
server_1  | [1]     at Runner.query (/srv/server/node_modules/knex/node_modules/bluebird/js/release/method.js:15:34)
server_1  | [1]     at /srv/server/node_modules/knex/lib/runner.js:65:21
server_1  | [1]     at tryCatcher (/srv/server/node_modules/knex/node_modules/bluebird/js/release/util.js:16:23)
server_1  | [1]     at /srv/server/node_modules/knex/node_modules/bluebird/js/release/using.js:185:26
server_1  | [1]     at tryCatcher (/srv/server/node_modules/knex/node_modules/bluebird/js/release/util.js:16:23)
server_1  | [1]     at Promise._settlePromiseFromHandler (/srv/server/node_modules/knex/node_modules/bluebird/js/release/promise.js:512:31)
server_1  | [1]   code: 'ER_TRUNCATED_WRONG_VALUE',
server_1  | [1]   errno: 1292,
server_1  | [1]   sqlMessage: 'Incorrect datetime value: \'2018-03-29T07:29:40Z\' for column \'created_at\' at row 1',
server_1  | [1]   sqlState: '22007',
server_1  | [1]   index: 0,
server_1  | [1]   sql: 'insert into `pushes` (`author`, `branch`, `created_at`, `message`, `sha`) values (\'simison\', \'master\', \'2018-03-29T07:29:40Z\', \'Security: add Monitor toggles and unify copy (#9084)\', \'b2523e5e39ca95c639acfaca88932df2421c0b30\')' }
```

This worked well :thumbsup:
```js
push.created_at = new Date(push.created_at);
```

File: .gitignore [hunk 2]
```diff
@@ -9,14 +9,16 @@ config/config.json
 /client/build
 
 # working dir
-/wp-calypso
+/server/repository
 
 # misc
 .DS_Store
+.env
```
Comment (simison) [inline .gitignore]: `docker-compose` picks this file automatically and sets env variables visible in containers.

File: .gitignore [hunk 3]
```diff
@@ -9,14 +9,16 @@ config/config.json
...
 .env.local
 .env.development.local
 .env.test.local
 .env.production.local
+data
```
Comment (simison) [inline .gitignore]: Contents of MySQL container get stored here.

File: server/builder.js [hunk 4]
```diff
@@ -131,15 +140,17 @@ async function analyzeBundle(push) {
 	} else {
 		// support for webpack < 4
 		newChunkGroups = [
-			{ sha, chunk: 'build', sibling: 'manifest' },
-			{ sha, chunk: 'build', sibling: 'vendor' },
+			{ sha: newStat.sha, chunk: 'build', sibling: 'manifest' },
+			{ sha: newStat.sha, chunk: 'build', sibling: 'vendor' },
```
Comment (jsnajdr) [inline server/builder.js]: I pushed a better fix for this bug in https://github.com/jsnajdr/icfy/commit/40a4fcfe1ecdb1ed399d2381f26ad73ccd3df448

There, I use `push.sha` property.

**Comment by jsnajdr** [comment]
> ER_DUP_ENTRY: Duplicate entry 'b2523e5e39ca95c639acfaca88932df2421c0b30' for key 'chunk_groups_sha_unique'

This was fixed in https://github.com/jsnajdr/icfy/commit/8731452b1f876398c0c2c815802c66543f9ede17

The columns in the `chunk_groups` table had incorrect `unique` constraint. You'll have to drop (`knex migrate:revert`) and recreate (`knex migrate:latest`) the `chunk_groups` table to make it work again.

File: server/queuer.js [hunk 5]
```diff
@@ -13,7 +17,7 @@ function sleep(ms) {
 }
 
 function printPush(push) {
-	return `${push.sha} at ${push.created_at} by ${push.author}: ${push.message}`;
+	return push ? `${push.sha} at ${push.created_at} by ${push.author}: ${push.message}` : '—';
```
Comment (simison) [inline server/queuer.js]: This fix is pretty much just hiding some other bug — maybe when the DB was completely empty I got undefined `push` somehow?

File: server/builder.js [hunk 6]
```diff
@@ -66,24 +67,26 @@ async function processPush(push) {
 	}
 
 	// update node_modules
-	await cmd('npm install');
+	await cmd('yarn install');
```
Comment (simison) [inline server/builder.js]: BTW http://blog.npmjs.org/post/171556855892/introducing-npm-ci-for-faster-more-reliable should be useful here.

File: server/queuer.js (hunk 5 above)
Comment (jsnajdr) [reply server/queuer.js]: That's very likely -- I vaguely remember ignoring special cases where there are no existing pushes. Now, after a few months, that neglect comes back to haunt me 😨

**Comment by simison** [comment]
This was fun but I think it's time to close it. :-)

**Pull request closed without merging**